### DISCUSSION TITLE
AST edges

### DISCUSSION THREAD

**Discussion by oxisto** [Fraunhofer-AISEC/cpg#427] [Ideas]
Moving the discussion of AST edges with @konradweiss to here, to follow up. Currently we face the following problems, we want to solve:

1. In the persisted (neo4j) graph it is currently **not easy** to follow the graph by *any* AST edge, because we have different edge types and a query currently must contain all the different AST types (which are in theory nonfinite)
2. Similarly, it is also **not easy** to retrieve an AST parent in the graph because of the different AST edge types
3. In the persisted (neo4j) graph it is currently **easy** to see the structure of a program because we have dedicated AST edge types (`LHS`, `RHS`, etc.)
4. In the in-memory graph, it is currently **impossible** to retrieve the AST parent of a particular node

**Comment by oxisto** [comment]
Before going into the detailed discussion, there are two main "possibilities" to tacke this problem.

# Solution 1: Specific AST edge types

With an added `ast` property set to `true` so that all ast-related edges can easily be queried using `MATCH p=()-[edge]-() WHERE edge.ast = true RETURN p` as well as specific AST edges like `MATCH p=()-[:LHS]-() RETURN p`. Is is quite easy to *see* the structure in the graph.

<img width="584" alt="Screenshot 2021-05-12 at 14 37 06" src="https://user-images.githubusercontent.com/12459061/117978108-d009ba80-b331-11eb-8ebe-11e4951cfc78.png">

# Solution 2: A generic AST edge

With an added `label` that contains the specific AST relation. Any AST edge can then easily be queried using `MATCH p=()-[:AST]-() RETURN p` and specific ones using `MATCH p=()-[edge:AST]-() WHERE edge.label = "LHS" RETURN p`. Is is not easy to *see* the structure in the graph.

<img width="512" alt="Screenshot 2021-05-12 at 12 48 55" src="https://user-images.githubusercontent.com/12459061/117978127-d435d800-b331-11eb-8b9a-7bd68bd0293f.png">

Any opinions about the solutions?

**Reply by konradweiss** [reply]
Solution 1

**Comment by Masrepus** [comment]
I think having some kind of AST property for edges is more convenient, as walking the AST ist mostly useful for queries, while the actual edge labels are more useful when looking at the graph. So not being able to directly see AST edges is probably less of a problem than not being able to see the actual labels.

**Reply by oxisto** [reply]
> I think having some kind of AST property for edges is more convenient, as walking the AST ist mostly useful for queries, while the actual edge labels are more useful when looking at the graph. So not being able to directly see AST edges is probably less of a problem than not being able to see the actual labels.

That would be solution 1 then, If I understood you correctly?

**Reply by Masrepus** [reply]
Yes 👍

**Reply by vfsrfs** [reply]
I agree with @Masrepus on having the AST as PropertyEdges is more convenient. The most common use case when looking at the edges is to know the label of an edge. If you then want to know if a particular edge is AST or not you can click on it, but otherwise you would have to click on a lot of edges if you want to see all the connections of a node when the labels are not visible.
